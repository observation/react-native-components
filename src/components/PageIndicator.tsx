import React from 'react'
import { StyleSheet, View } from 'react-native'

import Log from '../lib/Log'
import theme from '../styles/theme'

/** Maximum number of dots to display, should be odd */
const maximumNumberOfDots = 7

type Dot = 'small' | 'large' | 'current'

type Props = {
  currentPage: number
  numberOfPages: number
}

const PageIndicator = ({ currentPage, numberOfPages }: Props) => {
  Log.debug('PageIndicator', currentPage, numberOfPages)

  const getIndicators = () => {
    Log.debug('PageIndicator:getIndicators')

    if (numberOfPages <= 1) return []

    const numberOfDots = Math.min(numberOfPages, maximumNumberOfDots)
    const middleDot = (maximumNumberOfDots + 1) / 2
    const hasOverflowDots = numberOfPages > maximumNumberOfDots
    const hasLeftOverflowDot = hasOverflowDots && currentPage > middleDot
    const hasRightOverflowDot = hasOverflowDots && currentPage <= numberOfPages - middleDot
    const currentDot = (() => {
      Log.debug('PageIndicator:getIndicators')

      if (hasLeftOverflowDot && hasRightOverflowDot) return middleDot - 1
      if (hasLeftOverflowDot) return maximumNumberOfDots - 1 - (numberOfPages - currentPage)
      return currentPage - 1
    })()

    const indicators: Dot[] = Array(numberOfDots).fill('large')
    indicators[currentDot] = 'current'
    if (hasLeftOverflowDot) {
      indicators[0] = 'small'
    }
    if (hasRightOverflowDot) {
      indicators[maximumNumberOfDots - 1] = 'small'
    }
    return indicators
  }

  return (
    <View style={styles.containerStyle}>
      {getIndicators().map((indicator, index) => {
        switch (indicator) {
          case 'small':
            return <View key={index} style={[styles.dot, styles.smallDot]} />
          case 'large':
            return <View key={index} style={styles.dot} />
          case 'current':
            return <View key={index} style={[styles.dot, styles.currentDot]} />
        }
      })}
    </View>
  )
}

export default PageIndicator

const styles = StyleSheet.create({
  containerStyle: {
    flexDirection: 'row',
    alignSelf: 'center',
  },
  dot: {
    alignSelf: 'center',
    aspectRatio: 1,
    width: 8,
    margin: 4,
    ...theme.rounded,
    backgroundColor: theme.color.greyBase,
  },
  smallDot: {
    margin: 6,
    width: 4,
  },
  currentDot: {
    backgroundColor: theme.color.white,
  },
})
