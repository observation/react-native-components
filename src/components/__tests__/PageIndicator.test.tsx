import React from 'react'

import { render } from '@testing-library/react-native'

import PageIndicator from '../PageIndicator'

describe('PageIndicator', () => {
  describe('Rendering', () => {
    test('Page 1/1', () => {
      const { toJSON } = render(<PageIndicator currentPage={1} numberOfPages={1} />)
      expect(toJSON()).toMatchSnapshot()
    })
    test('Page 1/4', () => {
      const { toJSON } = render(<PageIndicator currentPage={1} numberOfPages={4} />)
      expect(toJSON()).toMatchSnapshot()
    })
    test('Page 2/4', () => {
      const { toJSON } = render(<PageIndicator currentPage={2} numberOfPages={4} />)
      expect(toJSON()).toMatchSnapshot()
    })
    test('Page 4/4', () => {
      const { toJSON } = render(<PageIndicator currentPage={4} numberOfPages={4} />)
      expect(toJSON()).toMatchSnapshot()
    })
    test('Page 1/12', () => {
      const { toJSON } = render(<PageIndicator currentPage={1} numberOfPages={12} />)
      expect(toJSON()).toMatchSnapshot()
    })
    test('Page 2/12', () => {
      const { toJSON } = render(<PageIndicator currentPage={2} numberOfPages={12} />)
      expect(toJSON()).toMatchSnapshot()
    })
    test('Page 5/12', () => {
      const { toJSON } = render(<PageIndicator currentPage={5} numberOfPages={12} />)
      expect(toJSON()).toMatchSnapshot()
    })
    test('Page 6/12', () => {
      const { toJSON } = render(<PageIndicator currentPage={6} numberOfPages={12} />)
      expect(toJSON()).toMatchSnapshot()
    })
    test('Page 7/12', () => {
      const { toJSON } = render(<PageIndicator currentPage={7} numberOfPages={12} />)
      expect(toJSON()).toMatchSnapshot()
    })
    test('Page 12/12', () => {
      const { toJSON } = render(<PageIndicator currentPage={12} numberOfPages={12} />)
      expect(toJSON()).toMatchSnapshot()
    })
  })
})
