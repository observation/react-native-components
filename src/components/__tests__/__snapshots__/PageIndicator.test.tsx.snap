// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageIndicator Rendering Page 1/1 1`] = `
<View
  style={
    {
      "alignSelf": "center",
      "flexDirection": "row",
    }
  }
/>
`;

exports[`PageIndicator Rendering Page 1/4 1`] = `
<View
  style={
    {
      "alignSelf": "center",
      "flexDirection": "row",
    }
  }
>
  <View
    style={
      [
        {
          "alignSelf": "center",
          "aspectRatio": 1,
          "backgroundColor": "#939393",
          "borderRadius": 4,
          "margin": 4,
          "overflow": "hidden",
          "width": 8,
        },
        {
          "backgroundColor": "#FFFFFF",
        },
      ]
    }
  />
  <View
    style={
      {
        "alignSelf": "center",
        "aspectRatio": 1,
        "backgroundColor": "#939393",
        "borderRadius": 4,
        "margin": 4,
        "overflow": "hidden",
        "width": 8,
      }
    }
  />
  <View
    style={
      {
        "alignSelf": "center",
        "aspectRatio": 1,
        "backgroundColor": "#939393",
        "borderRadius": 4,
        "margin": 4,
        "overflow": "hidden",
        "width": 8,
      }
    }
  />
  <View
    style={
      {
        "alignSelf": "center",
        "aspectRatio": 1,
        "backgroundColor": "#939393",
        "borderRadius": 4,
        "margin": 4,
        "overflow": "hidden",
        "width": 8,
      }
    }
  />
</View>
`;

exports[`PageIndicator Rendering Page 1/12 1`] = `
<View
  style={
    {
      "alignSelf": "center",
      "flexDirection": "row",
    }
  }
>
  <View
    style={
      [
        {
          "alignSelf": "center",
          "aspectRatio": 1,
          "backgroundColor": "#939393",
          "borderRadius": 4,
          "margin": 4,
          "overflow": "hidden",
          "width": 8,
        },
        {
          "backgroundColor": "#FFFFFF",
        },
      ]
    }
  />
  <View
    style={
      {
        "alignSelf": "center",
        "aspectRatio": 1,
        "backgroundColor": "#939393",
        "borderRadius": 4,
        "margin": 4,
        "overflow": "hidden",
        "width": 8,
      }
    }
  />
  <View
    style={
      {
        "alignSelf": "center",
        "aspectRatio": 1,
        "backgroundColor": "#939393",
        "borderRadius": 4,
        "margin": 4,
        "overflow": "hidden",
        "width": 8,
      }
    }
  />
  <View
    style={
      {
        "alignSelf": "center",
        "aspectRatio": 1,
        "backgroundColor": "#939393",
        "borderRadius": 4,
        "margin": 4,
        "overflow": "hidden",
        "width": 8,
      }
    }
  />
  <View
    style={
      {
        "alignSelf": "center",
        "aspectRatio": 1,
        "backgroundColor": "#939393",
        "borderRadius": 4,
        "margin": 4,
        "overflow": "hidden",
        "width": 8,
      }
    }
  />
  <View
    style={
      {
        "alignSelf": "center",
        "aspectRatio": 1,
        "backgroundColor": "#939393",
        "borderRadius": 4,
        "margin": 4,
        "overflow": "hidden",
        "width": 8,
      }
    }
  />
  <View
    style={
      [
        {
          "alignSelf": "center",
          "aspectRatio": 1,
          "backgroundColor": "#939393",
          "borderRadius": 4,
          "margin": 4,
          "overflow": "hidden",
          "width": 8,
        },
        {
          "margin": 6,
          "width": 4,
        },
      ]
    }
  />
</View>
`;

exports[`PageIndicator Rendering Page 2/4 1`] = `
<View
  style={
    {
      "alignSelf": "center",
      "flexDirection": "row",
    }
  }
>
  <View
    style={
      {
        "alignSelf": "center",
        "aspectRatio": 1,
        "backgroundColor": "#939393",
        "borderRadius": 4,
        "margin": 4,
        "overflow": "hidden",
        "width": 8,
      }
    }
  />
  <View
    style={
      [
        {
          "alignSelf": "center",
          "aspectRatio": 1,
          "backgroundColor": "#939393",
          "borderRadius": 4,
          "margin": 4,
          "overflow": "hidden",
          "width": 8,
        },
        {
          "backgroundColor": "#FFFFFF",
        },
      ]
    }
  />
  <View
    style={
      {
        "alignSelf": "center",
        "aspectRatio": 1,
        "backgroundColor": "#939393",
        "borderRadius": 4,
        "margin": 4,
        "overflow": "hidden",
        "width": 8,
      }
    }
  />
  <View
    style={
      {
        "alignSelf": "center",
        "aspectRatio": 1,
        "backgroundColor": "#939393",
        "borderRadius": 4,
        "margin": 4,
        "overflow": "hidden",
        "width": 8,
      }
    }
  />
</View>
`;

exports[`PageIndicator Rendering Page 2/12 1`] = `
<View
  style={
    {
      "alignSelf": "center",
      "flexDirection": "row",
    }
  }
>
  <View
    style={
      {
        "alignSelf": "center",
        "aspectRatio": 1,
        "backgroundColor": "#939393",
        "borderRadius": 4,
        "margin": 4,
        "overflow": "hidden",
        "width": 8,
      }
    }
  />
  <View
    style={
      [
        {
          "alignSelf": "center",
          "aspectRatio": 1,
          "backgroundColor": "#939393",
          "borderRadius": 4,
          "margin": 4,
          "overflow": "hidden",
          "width": 8,
        },
        {
          "backgroundColor": "#FFFFFF",
        },
      ]
    }
  />
  <View
    style={
      {
        "alignSelf": "center",
        "aspectRatio": 1,
        "backgroundColor": "#939393",
        "borderRadius": 4,
        "margin": 4,
        "overflow": "hidden",
        "width": 8,
      }
    }
  />
  <View
    style={
      {
        "alignSelf": "center",
        "aspectRatio": 1,
        "backgroundColor": "#939393",
        "borderRadius": 4,
        "margin": 4,
        "overflow": "hidden",
        "width": 8,
      }
    }
  />
  <View
    style={
      {
        "alignSelf": "center",
        "aspectRatio": 1,
        "backgroundColor": "#939393",
        "borderRadius": 4,
        "margin": 4,
        "overflow": "hidden",
        "width": 8,
      }
    }
  />
  <View
    style={
      {
        "alignSelf": "center",
        "aspectRatio": 1,
        "backgroundColor": "#939393",
        "borderRadius": 4,
        "margin": 4,
        "overflow": "hidden",
        "width": 8,
      }
    }
  />
  <View
    style={
      [
        {
          "alignSelf": "center",
          "aspectRatio": 1,
          "backgroundColor": "#939393",
          "borderRadius": 4,
          "margin": 4,
          "overflow": "hidden",
          "width": 8,
        },
        {
          "margin": 6,
          "width": 4,
        },
      ]
    }
  />
</View>
`;

exports[`PageIndicator Rendering Page 4/4 1`] = `
<View
  style={
    {
      "alignSelf": "center",
      "flexDirection": "row",
    }
  }
>
  <View
    style={
      {
        "alignSelf": "center",
        "aspectRatio": 1,
        "backgroundColor": "#939393",
        "borderRadius": 4,
        "margin": 4,
        "overflow": "hidden",
        "width": 8,
      }
    }
  />
  <View
    style={
      {
        "alignSelf": "center",
        "aspectRatio": 1,
        "backgroundColor": "#939393",
        "borderRadius": 4,
        "margin": 4,
        "overflow": "hidden",
        "width": 8,
      }
    }
  />
  <View
    style={
      {
        "alignSelf": "center",
        "aspectRatio": 1,
        "backgroundColor": "#939393",
        "borderRadius": 4,
        "margin": 4,
        "overflow": "hidden",
        "width": 8,
      }
    }
  />
  <View
    style={
      [
        {
          "alignSelf": "center",
          "aspectRatio": 1,
          "backgroundColor": "#939393",
          "borderRadius": 4,
          "margin": 4,
          "overflow": "hidden",
          "width": 8,
        },
        {
          "backgroundColor": "#FFFFFF",
        },
      ]
    }
  />
</View>
`;

exports[`PageIndicator Rendering Page 5/12 1`] = `
<View
  style={
    {
      "alignSelf": "center",
      "flexDirection": "row",
    }
  }
>
  <View
    style={
      [
        {
          "alignSelf": "center",
          "aspectRatio": 1,
          "backgroundColor": "#939393",
          "borderRadius": 4,
          "margin": 4,
          "overflow": "hidden",
          "width": 8,
        },
        {
          "margin": 6,
          "width": 4,
        },
      ]
    }
  />
  <View
    style={
      {
        "alignSelf": "center",
        "aspectRatio": 1,
        "backgroundColor": "#939393",
        "borderRadius": 4,
        "margin": 4,
        "overflow": "hidden",
        "width": 8,
      }
    }
  />
  <View
    style={
      {
        "alignSelf": "center",
        "aspectRatio": 1,
        "backgroundColor": "#939393",
        "borderRadius": 4,
        "margin": 4,
        "overflow": "hidden",
        "width": 8,
      }
    }
  />
  <View
    style={
      [
        {
          "alignSelf": "center",
          "aspectRatio": 1,
          "backgroundColor": "#939393",
          "borderRadius": 4,
          "margin": 4,
          "overflow": "hidden",
          "width": 8,
        },
        {
          "backgroundColor": "#FFFFFF",
        },
      ]
    }
  />
  <View
    style={
      {
        "alignSelf": "center",
        "aspectRatio": 1,
        "backgroundColor": "#939393",
        "borderRadius": 4,
        "margin": 4,
        "overflow": "hidden",
        "width": 8,
      }
    }
  />
  <View
    style={
      {
        "alignSelf": "center",
        "aspectRatio": 1,
        "backgroundColor": "#939393",
        "borderRadius": 4,
        "margin": 4,
        "overflow": "hidden",
        "width": 8,
      }
    }
  />
  <View
    style={
      [
        {
          "alignSelf": "center",
          "aspectRatio": 1,
          "backgroundColor": "#939393",
          "borderRadius": 4,
          "margin": 4,
          "overflow": "hidden",
          "width": 8,
        },
        {
          "margin": 6,
          "width": 4,
        },
      ]
    }
  />
</View>
`;

exports[`PageIndicator Rendering Page 6/12 1`] = `
<View
  style={
    {
      "alignSelf": "center",
      "flexDirection": "row",
    }
  }
>
  <View
    style={
      [
        {
          "alignSelf": "center",
          "aspectRatio": 1,
          "backgroundColor": "#939393",
          "borderRadius": 4,
          "margin": 4,
          "overflow": "hidden",
          "width": 8,
        },
        {
          "margin": 6,
          "width": 4,
        },
      ]
    }
  />
  <View
    style={
      {
        "alignSelf": "center",
        "aspectRatio": 1,
        "backgroundColor": "#939393",
        "borderRadius": 4,
        "margin": 4,
        "overflow": "hidden",
        "width": 8,
      }
    }
  />
  <View
    style={
      {
        "alignSelf": "center",
        "aspectRatio": 1,
        "backgroundColor": "#939393",
        "borderRadius": 4,
        "margin": 4,
        "overflow": "hidden",
        "width": 8,
      }
    }
  />
  <View
    style={
      [
        {
          "alignSelf": "center",
          "aspectRatio": 1,
          "backgroundColor": "#939393",
          "borderRadius": 4,
          "margin": 4,
          "overflow": "hidden",
          "width": 8,
        },
        {
          "backgroundColor": "#FFFFFF",
        },
      ]
    }
  />
  <View
    style={
      {
        "alignSelf": "center",
        "aspectRatio": 1,
        "backgroundColor": "#939393",
        "borderRadius": 4,
        "margin": 4,
        "overflow": "hidden",
        "width": 8,
      }
    }
  />
  <View
    style={
      {
        "alignSelf": "center",
        "aspectRatio": 1,
        "backgroundColor": "#939393",
        "borderRadius": 4,
        "margin": 4,
        "overflow": "hidden",
        "width": 8,
      }
    }
  />
  <View
    style={
      [
        {
          "alignSelf": "center",
          "aspectRatio": 1,
          "backgroundColor": "#939393",
          "borderRadius": 4,
          "margin": 4,
          "overflow": "hidden",
          "width": 8,
        },
        {
          "margin": 6,
          "width": 4,
        },
      ]
    }
  />
</View>
`;

exports[`PageIndicator Rendering Page 7/12 1`] = `
<View
  style={
    {
      "alignSelf": "center",
      "flexDirection": "row",
    }
  }
>
  <View
    style={
      [
        {
          "alignSelf": "center",
          "aspectRatio": 1,
          "backgroundColor": "#939393",
          "borderRadius": 4,
          "margin": 4,
          "overflow": "hidden",
          "width": 8,
        },
        {
          "margin": 6,
          "width": 4,
        },
      ]
    }
  />
  <View
    style={
      {
        "alignSelf": "center",
        "aspectRatio": 1,
        "backgroundColor": "#939393",
        "borderRadius": 4,
        "margin": 4,
        "overflow": "hidden",
        "width": 8,
      }
    }
  />
  <View
    style={
      {
        "alignSelf": "center",
        "aspectRatio": 1,
        "backgroundColor": "#939393",
        "borderRadius": 4,
        "margin": 4,
        "overflow": "hidden",
        "width": 8,
      }
    }
  />
  <View
    style={
      [
        {
          "alignSelf": "center",
          "aspectRatio": 1,
          "backgroundColor": "#939393",
          "borderRadius": 4,
          "margin": 4,
          "overflow": "hidden",
          "width": 8,
        },
        {
          "backgroundColor": "#FFFFFF",
        },
      ]
    }
  />
  <View
    style={
      {
        "alignSelf": "center",
        "aspectRatio": 1,
        "backgroundColor": "#939393",
        "borderRadius": 4,
        "margin": 4,
        "overflow": "hidden",
        "width": 8,
      }
    }
  />
  <View
    style={
      {
        "alignSelf": "center",
        "aspectRatio": 1,
        "backgroundColor": "#939393",
        "borderRadius": 4,
        "margin": 4,
        "overflow": "hidden",
        "width": 8,
      }
    }
  />
  <View
    style={
      [
        {
          "alignSelf": "center",
          "aspectRatio": 1,
          "backgroundColor": "#939393",
          "borderRadius": 4,
          "margin": 4,
          "overflow": "hidden",
          "width": 8,
        },
        {
          "margin": 6,
          "width": 4,
        },
      ]
    }
  />
</View>
`;

exports[`PageIndicator Rendering Page 12/12 1`] = `
<View
  style={
    {
      "alignSelf": "center",
      "flexDirection": "row",
    }
  }
>
  <View
    style={
      [
        {
          "alignSelf": "center",
          "aspectRatio": 1,
          "backgroundColor": "#939393",
          "borderRadius": 4,
          "margin": 4,
          "overflow": "hidden",
          "width": 8,
        },
        {
          "margin": 6,
          "width": 4,
        },
      ]
    }
  />
  <View
    style={
      {
        "alignSelf": "center",
        "aspectRatio": 1,
        "backgroundColor": "#939393",
        "borderRadius": 4,
        "margin": 4,
        "overflow": "hidden",
        "width": 8,
      }
    }
  />
  <View
    style={
      {
        "alignSelf": "center",
        "aspectRatio": 1,
        "backgroundColor": "#939393",
        "borderRadius": 4,
        "margin": 4,
        "overflow": "hidden",
        "width": 8,
      }
    }
  />
  <View
    style={
      {
        "alignSelf": "center",
        "aspectRatio": 1,
        "backgroundColor": "#939393",
        "borderRadius": 4,
        "margin": 4,
        "overflow": "hidden",
        "width": 8,
      }
    }
  />
  <View
    style={
      {
        "alignSelf": "center",
        "aspectRatio": 1,
        "backgroundColor": "#939393",
        "borderRadius": 4,
        "margin": 4,
        "overflow": "hidden",
        "width": 8,
      }
    }
  />
  <View
    style={
      {
        "alignSelf": "center",
        "aspectRatio": 1,
        "backgroundColor": "#939393",
        "borderRadius": 4,
        "margin": 4,
        "overflow": "hidden",
        "width": 8,
      }
    }
  />
  <View
    style={
      [
        {
          "alignSelf": "center",
          "aspectRatio": 1,
          "backgroundColor": "#939393",
          "borderRadius": 4,
          "margin": 4,
          "overflow": "hidden",
          "width": 8,
        },
        {
          "backgroundColor": "#FFFFFF",
        },
      ]
    }
  />
</View>
`;
